server/src/main/java/org/elasticsearch/transport/ClusterConnectionManager.java

```diff
@@ -103,7 +107,11 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
             return;
         }
 
-        if (connectedNodes.containsKey(node)) {
+        ConnectedNodeRefCounter connectedNodeRefCounter = connectedNodes.get(node);
+        if (connectedNodeRefCounter != null) {
+            if (connectedNodeRefCounter.isDeleting) {
+                connectedNodeRefCounter.incRef();
```

Comment:
Unfortunately this will leak: we might call `connectToNode` multiple times while the `isDeleting` flag is set (e.g. handling multiple joins from the same node, or else an old `PeerFinder` might still be active). We need to make sure that every `connectToNode` call explicitly releases any references that it obtained.